source/opt/unify_const.cpp

```diff
@@ -0,0 +1,169 @@
+
+#include <unordered_set>
+
+#include "def_use_manager.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// Using a trie node to store encountered constants and look up later with a
+// given constant defining instruction's opcode, result type id and operands.
+// Each ResultIdTrieNode stores a result id, if it does represent an
+// encountered constant. ResultIdTrieNode also guide the lookup process by
+// holds the map toward the next ResultIdTrieNode to check with.
+struct ResultIdTrieNode {
+  ResultIdTrieNode() : result_id(0), next() {}
+  uint32_t result_id;
+  std::unordered_map<uint64_t, std::unique_ptr<ResultIdTrieNode>> next;
+};
+
+// A helper function to extract the operand's raw value in 64-bit unsigned
+// integers. Note that this does not support operands longer than 64-bit.
+uint64_t GetOperandValueInUint64(const ir::Operand& operand) {
+  assert(operand.words.size() < 3 &&
+         "Only support operands no longer than 64 bits");
+  uint64_t result = 0;
+  for (size_t i = 0; i < operand.words.size(); i++) {
+    result |= operand.words[i] << (i * 32);
+  }
+  return result;
+}
+
+// A helper function to ease trie node looking up. Returns a trie node pointer
+// pointing to the node identified by the given key in the given trie node map.
+// If there is no such node for the given key, a new trie node will be built
+// and then a pointer to the new built node will be returned.
+ResultIdTrieNode* GetOrCreateIteratorFor(
+    uint64_t key,
+    std::unordered_map<uint64_t, std::unique_ptr<ResultIdTrieNode>>* m) {
+  auto iter = m->find(key);
+  if (iter == m->end()) {
+    // insert new entry
+    (*m)[key] = std::unique_ptr<ResultIdTrieNode>(new ResultIdTrieNode);
+    iter = m->find(key);
+  }
+  return iter->second.get();
+}
+
+// A helper function to construct the keys used to identify a constant. The
+// keys are consisent of 1) opcode, 2) result type id, 3) all other operands.
+// Constants with same keys are guaranteed to be exactly the same.
+std::vector<uint64_t> GetLookUpKeys(const ir::Instruction& inst) {
+  std::vector<uint64_t> keys;
+  // Need to use the opcode, otherwise there might be a conflict with the
+  // following case when <op>'s binary value equals xx's id:
+  //  OpSpecConstantOp tt <op> yy zz
+  //  OpSpecConstantComposite tt xx yy zz;
+  keys.push_back(static_cast<uint64_t>(inst.opcode()));
+  for (uint32_t i = 0; i < inst.NumOperands(); i++) {
+    if (inst.GetOperand(i).type == SPV_OPERAND_TYPE_RESULT_ID) continue;
+    keys.push_back(GetOperandValueInUint64(inst.GetOperand(i)));
+  }
+  return keys;
+}
+
+// Tries to find the same constant defining instruction of the given |inst| in
+// the trie pointed by the given trie node |node|. If a same instruction is
+// found in the trie, returns its result id. Otherwise, creates an trie node
+// for the given instruction then returns its result id.
+uint32_t LookupOrInsert(ResultIdTrieNode* node, const ir::Instruction& inst) {
```

Comment:
It looks like this function should be a member method of `class ResultIdTrieNode`.